sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -360,7 +364,15 @@ class Dataset[T] private[sql](
    * @since 1.6.0
    */
   def select[U1: Encoder](c1: TypedColumn[T, U1]): Dataset[U1] = {
-    new Dataset[U1](sqlContext, Project(Alias(withEncoder(c1).expr, "_1")() :: Nil, logicalPlan))
+    // We use an unbound encoder since the expression will make up its own schema.
+    // TODO: This probably doesn't work if we are relying on reordering of the input class fields.
```

Comment:
Yeah, it is.  Actually, this comment is stale and there is a [test now](https://github.com/apache/spark/pull/9673/files#diff-bf07c7708b6eceda37d9c5c8b5be0d54R157) to show that it works.